tests/transformers/test_transformers_model_export.py

```diff
@@ -2501,163 +2484,119 @@ def test_instructional_pipeline_with_prompt_in_output(model_path):
-    # NB: This test validates pre-processing via converting the sounds file into the
-    # appropriate bitrate encoding rate and casting to a numpy array. Other tests validate
-    # the 'raw' file input of bytes.
+def read_audio_data(format: str):
+    datasets_path = pathlib.Path(__file__).resolve().parent.parent.joinpath("datasets")
+    wav_file_path = datasets_path.joinpath("apollo11_launch.wav")
+    if format == "float":
+        audio, _ = librosa.load(wav_file_path, sr=16000)
+        return audio
+    elif format == "bytes":
+        return wav_file_path.read_bytes()
+    elif format == "file":
+        return wav_file_path.as_posix()
+    else:
+        raise ValueError(f"Invalid format: {format}")
 
-    model_config = {
-        "return_timestamps": "word",
-        "chunk_length_s": 20,
-        "stride_length_s": [5, 3],
-    }
 
-    signature = infer_signature(
-        sound_file_for_test,
-        mlflow.transformers.generate_signature_output(whisper_pipeline, sound_file_for_test),
-    )
+@pytest.mark.parametrize("input_format", ["float", "bytes", "file"])
+@pytest.mark.parametrize("with_input_example", [True, False])
+def test_whisper_model_predict(model_path, whisper_pipeline, input_format, with_input_example):
+    if input_format == "float" and not with_input_example:
+        pytest.skip("If the input format is float, the default signature must be overridden.")
 
+    audio = read_audio_data(input_format)
     mlflow.transformers.save_model(
         transformers_model=whisper_pipeline,
         path=model_path,
-        model_config=model_config,
-        signature=signature,
+        input_example=audio if with_input_example else None,
+        save_pretrained=False,
     )
 
+    # 1. Single prediction with native transformer pipeline
     loaded_pipeline = mlflow.transformers.load_model(model_path)
-
-    transcription = loaded_pipeline(sound_file_for_test, **model_config)
+    transcription = loaded_pipeline(audio)
     assert transcription["text"].startswith(" 30")
+    # strip the leading space
+    expected_text = transcription["text"].strip()
 
+    # 2. Single prediction with Pyfunc
     loaded_pyfunc = mlflow.pyfunc.load_model(model_path)
+    pyfunc_transcription = loaded_pyfunc.predict(audio)[0]
+    assert pyfunc_transcription == expected_text
 
-    pyfunc_transcription = json.loads(loaded_pyfunc.predict(sound_file_for_test)[0])
-
-    assert transcription["text"] == pyfunc_transcription["text"]
-    # Due to the choice of using tuples within the return type, equivalency validation for the
-    # "chunks" values is not explicitly equivalent since tuples are cast to lists when json
-    # serialized.
-    assert transcription["chunks"][0]["text"] == pyfunc_transcription["chunks"][0]["text"]
+    # 3. Batch prediction with Pyfunc. Float input format is not supported for batch prediction,
+    # because our signature framework doesn't support a list of numpy array.
+    if input_format != "float":
```

Comment:
note: Supporting batch inference for numpy array input might be tricky. Leaving it as unsupported as it's not required for the current customer request.

Reply:
The input signature for a collection of numpy arrays would be a tensor, but the transformers audio implementation doesn't support tensor input. Should we provide an exception message to alert users who might try this to suggest to use a different input format for batch inference of audio transcriptions?

Reply:
> Should we provide an exception message to alert users who might try this to suggest to use a different input format for batch inference of audio transcriptions?

Hmm the tricky thing is that the exception happens when enforcing schema in pyfunc, before entering to the Transformers wrapper.

> E               mlflow.exceptions.MlflowException: Failed to enforce schema of data '[array([-0.01484766, -0.01701006, -0.01464985, ...,  0.12725656,
E                       0.2055437 ,  0.14151388], dtype=float32), array([-0.01484766, -0.01701006, -0.01464985, ...,  0.12725656,
E                       0.2055437 ,  0.14151388], dtype=float32)]' with schema '[Tensor('float32', (-1,))]'. Error: This model contains a tensor-based model signature with no input names, which suggests a numpy array input or a pandas dataframe input with proper column values, but an input of type <class 'list'> was found.

To translate this exception to the special message, we need to check if transformer model is audio pipeline or not within PythonModel [here](https://github.com/mlflow/mlflow/blob/2bbf9e6d3826a7ea153477b609cfa2a8b513a3a7/mlflow/pyfunc/__init__.py#L699). It's doable, but I'm not very comfortable with leaking flavor logic to general PythonModel implementation.

My opinion is that, given Transformers model itself doesn't support batch inference for float array, there shouldn't be many users who think it's possible in MLflow. Then we don't have to worry too much, at least until getting explicit question from someone who tries that. Alternatively, we could add a quick note in the documentation, but it's might bit challenging to make it easy to reach. What do you think?

Reply:
This logic does not belong in pyfunc. Agree with the original design here due to overall complexity and it not really belonging in the parsing logic that has nothing to do with schema enforcement.